Add JobJobEventsChildrenSummary endpoint

fosterseth opened the pull request:
<!--- changelog-entry
# Fill in 'msg' below to have an entry automatically added to the next release changelog.
# Leaving 'msg' blank will not generate a changelog entry for this PR.
# Please ensure this is a simple (and readable) one-line string.
---
msg: ""
-->

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Adds a special endpoint for the purposes of optimizing the Job output page in the UI

Events are hierarchical

```

E1
- E2
- - E3
- - E4
- E5

```

- E1 has 4 total children
- E2 has 2 total children
- E2 and E5 are siblings

The UI needs to know how many children (and grandchildren, etc.) belong to a particular event. Currently the UI does a couple of API calls to get this information, _per_ nested event branch. These extraneous networks calls were leading to issues

https://github.com/ansible/awx/issues/11765#issuecomment-1066985723

This new endpoint is job specific, and returns a json object where each key is the event counter, and the value is the total number of children under that event. The idea is that the UI can get this data once, then use it as a lookup table any time it needs to know the number of children that belong to an event.

for the above event structure it would be
```json
{
1: 4,
2: 2,
}
```
Note, events with no children are not included. Also, it only returns data if the job is in a completed state.


<!---
If you are fixing an existing issue, please include "related #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - API

##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```
awx: 20.0.2.dev101+g0d3af6ec4a
```

AlanCoding commented on awx/api/views/__init__.py:
```diff
@@ -3824,6 +3824,82 @@ def get_queryset(self):
+    Only events with children are included in the output.
+
+    e.g. Demo Job Template job
+    given:
```
By "given", you just mean that this is the event data for the Demo Job Template?

fosterseth replied:
exactly

AlanCoding commented on awx/api/urls/job.py:
```diff
@@ -27,6 +28,7 @@
     re_path(r'^(?P<pk>[0-9]+)/create_schedule/$', JobCreateSchedule.as_view(), name='job_create_schedule'),
     re_path(r'^(?P<pk>[0-9]+)/job_host_summaries/$', JobJobHostSummariesList.as_view(), name='job_job_host_summaries_list'),
     re_path(r'^(?P<pk>[0-9]+)/job_events/$', JobJobEventsList.as_view(), name='job_job_events_list'),
+    re_path(r'^(?P<pk>[0-9]+)/job_events/children_summary/$', JobJobEventsChildrenSummary.as_view(), name='job_job_event_children_summary'),
```
Does the UI really need this specific to jobs, or would it be easier to apply it to all unified jobs, including project updates and ad-hoc commands? Your logic (like `job.get_event_queryset()`) could mostly work for all job types. Best to ask these questions before making the schema change.

fosterseth replied:
@keithjgrant is this only specific to Jobs? Would it be best to have children_summary endpoint for other types of jobs?

ProjectUpdateEvent does not have a parent_uuid field

keithjgrant replied:
Currently the UI only shows the collapsible view for playbook jobs. If we want that functionality for any other job types, I wouldn't be opposed (assuming it makes sense to), but they would also need their parent_uuid's populated, which we don't have now

AlanCoding commented on awx/api/views/__init__.py:
```diff
@@ -3824,6 +3824,82 @@ def get_queryset(self):
+                continue
+            map_uuid_counter[uuid] = counter
+            if not puuid:
+                continue
```
For these cases, it would be better to filter out empty/null values when creating the original queryset. You're already at risk of performance problems, and that _could_ help.

fosterseth replied:
good call

AlanCoding commented on awx/api/views/__init__.py:
```diff
@@ -3824,6 +3824,82 @@ def get_queryset(self):
+        map_counter_children_tally = dict()
+
+        def update_grandparents(uuid):
+            # recursive function that updates the tally for grandparent events
```
I'm not a die-hard purist against recursive methods, but in this case it looks like it would be pretty simple to re-write this as a nested loop and avoid the recursion.

fosterseth replied:
changed this to a while loop

AlanCoding commented on awx/api/views/__init__.py:
```diff
@@ -3824,6 +3824,82 @@ def get_queryset(self):
+        "1": 8,
+        "2": 6,
+        "3": 2,
+        "6": 2
```
Even if this method can run in a reasonable amount of time for a high-output job, the _number of keys_ is going to make it difficult to return in the response. If there are a million events, will it even be able to _transmit_ the response before the uWSGI worker dies? As a whole-job endpoint, I find it doubtful that this endpoint will survive the test of real users.

AlanCoding replied:
Then again, I get where you're coming from excluding events with _no_ children. Assuming the multi-million or multi-billion events are because of number of hosts... maybe. How do loops figure into this? If I run one host but loop a task a million times, what fraction of the events have children?

fosterseth replied:
yeah this is where things aren't clear me to me. The output here only includes events with children, which I am hoping is a small enough subset to reliably do this kind of thing. We may need to find a realistic, high output job to test this on.

AlanCoding replied:
Running chatty_tasks with 1 host

```
In [6]: parents = set(event.parent_uuid for event in events)

In [7]: len(events)
Out[7]: 56

In [8]: len([event for event in events if event.uuid in parents])
Out[8]: 3
```

So seems fairly valid to cut out the bulk of data from loops.

fosterseth committed:
Add JobJobEventsChildrenSummary endpoint

- returns a special view to output the total number of children (and
grandchildren) events for all parents events for a job
value is the number of total children of that event
- intended to be consumed by the UI, as an efficient way to get the
number of children for a particular event
- see api/templates/api/job_job_events_children_summary.md for more info

AlanCoding commented on awx/api/views/__init__.py:
```diff
@@ -3824,6 +3824,75 @@ def get_queryset(self):
+    The UI also needs to know the order of the event (0 based index), which
+    usually matches the counter, but not always.
+    This view returns a JSON object where the key is the event counter, and the value
+    includes the number of children (and grandchildren) events.
```
value includes something else now

AlanCoding commented on awx/api/views/__init__.py:
```diff
@@ -3824,6 +3824,75 @@ def get_queryset(self):
+        "3": [2, 2],
+        "6": [5, 2],
+        }
+    counter 1 is event 0, and has 8 children
```
what does "is event 0" mean? that 0 is the primary key?

fosterseth commented on awx/api/views/__init__.py:
```diff
@@ -3826,6 +3826,174 @@ def get_queryset(self):
 
 
+class JobJobEventsChildrenSummary(APIView):
+    '''
```
if people feel this doc string is too long, we can add this as an entry in docs/

AlanCoding commented on awx/api/views/__init__.py:
```diff
@@ -3826,6 +3826,168 @@ def get_queryset(self):
+    def get(self, request, **kwargs):
+        job = get_object_or_404(models.Job, pk=kwargs['pk'])
+        if not job.event_processing_finished:
+            return Response({})
```
If this won't run when event processing isn't finished, then why do we need to bother with the rowNumber? It should be the same as counter minus 1.

keithjgrant replied:
@AlanCoding does that mean there are never gaps in the counter values once processing is done? If that's the case, we can sort that out in the UI.

It also means I could probably look for ways to simplify the UI logic, as it does extra work to handle cases where there's a counter gap. Can there still be a gap while the job is running/processing?

AlanCoding replied:
There can't reasonably be a gap when `job.event_processing_finished` is True. Consider the logic from that.

https://github.com/ansible/awx/blob/6c1adade250020fcf900ecfc6d33e6f8274c8e0a/awx/main/models/unified_jobs.py#L1022

The way this works:
 - Control process receives output from receptor, and it does minimal processing of this and submits the data to receptor
   - This process counts the number of events that it sends, and saves that to `emitted_events`
 - Callback receiver pulls the data off the redis queue and then saves it to database

If `event_processing_finished` never transitions to `True`, then that's an API bug. Doesn't matter what type of job or what it does. This is accounting - the callback receiver should always save the same number of events that the control process submitted.

If the job is actually _running_ then this endpoint returns `{}` anyway

fosterseth replied:
there are 3 numbers to consider here

- event_qs.count()
- emitted_events
- the `counter` field on the event itself (from ansible-runner)

the first 2 are pretty much guaranteed to match if event_processing_finished==true, as Alan outlined above.

What about `counter` field on the event itself? if there are 100 events, are we guaranteed that we have events with counters 1 through 100? Looking at code I don't see why not. I'm wondering if the previous reports of missing counters was because of not checking whether events had been fully processed or not.

We could make the decision to rely on `counter`, simplifying this endpoint a bit. We could just throw an error or warning if this code finds some missing counters.

AlanCoding commented on awx/api/views/__init__.py:
```diff
@@ -3826,6 +3826,168 @@ def get_queryset(self):
+    def get(self, request, **kwargs):
+        job = get_object_or_404(models.Job, pk=kwargs['pk'])
+        if not job.event_processing_finished:
+            return Response({})
```
Right after this line, I think you should go ahead and write a conditional to give a 400 with too many events. You don't need to load anything, use `job.emitted_events`, already in memory. If that number is greater than your limit, then bail.

How do you know what the limit is? Use the firehose script or whatever you want to get the time taken to load some events. Divide to get seconds per event, and then use that to find out how many events will take 10 seconds to load and make that the limit. This is simple enough that it would be best to do that now and avoid more costly iteration in the future.

rebeccahhh reviewed (approved):
I like these updates a lot!
